OpenAPI Generation

bjhham opened the pull request:
This is a proposal for an annotation-based OpenAPI specification generator.

Work item: https://youtrack.jetbrains.com/issue/KTOR-8316/
Template project: https://github.com/bjhham/open-api-eap-template 

Our goal is to introduce a simple and unobtrusive way to provide the necessary details for exposing OpenAPI specifications from Ktor applications.

[Rendered preview](https://github.com/ktorio/ktor-klip/blob/openapi/proposals/0004-open-api.md)

bjhham committed:
Added OpenAPI generation proposal

bjhham committed:
Added proposal using KDoc approach

bjhham committed:
A little more information on the proposed solution

bjhham committed:
Add gradle task example

bjhham committed:
Add details

bjhham committed:
Update index

bjhham committed:
Add extensibility

bjhham committed:
Fix header; add example

nomisRev commented on proposals/0004-open-api.md:
```diff
@@ -0,0 +1,322 @@
+
+To address this requirement, we indent to supplement the path information with a KDoc-like annotation API that can be read by our Gradle plugin.
+
+## Annotation API
+[annotation-api]: #annotation-api
```
Can we give this a nice name like "KDoc-umenting API" 😁

bjhham replied:
I should probably give it some kinda different name since "annotation" is stolen by annotation classes...

bjhham commented on proposals/0004-open-api.md:
````diff
@@ -0,0 +1,322 @@
+Where the `OpenAPISource` argument is a functional interface:
+
+```kotlin
+fun interface OpenAPISource {
+    fun generate(application: Application): OpenAPIFragment
+}
+```
````
Thank you, rabbit.  I will take this into consideration.

BoD commented on proposals/0004-open-api.md:
````diff
@@ -0,0 +1,322 @@
+}
+```
+
+As shown in the example, we intend to inject the missing path information using the KDoc comment syntax.  Developers will be supported by IDE tooling to resolve code references in the comments, and it will prevent the need to modify any existing routes in the current routing API.
````
Why not using regular annotations, as opposed to special markup inside comments? It seems to me that it would allow other tools to more easily benefit from the added information.

e5l replied:
regular annotations looks fine for me if they are used not for code generation, but metadata

bjhham replied:
I think because we have to target expressions in the code, as opposed to classes/functions, switching to annotations wouldn't provide a significant benefit for external tooling.  For this reason, we're letting aesthetics drive the choice, which is why we're looking at KDoc comments.

BoD replied:
I see.

I still feel this would be kind of a missed opportunity. It's hard to predict what kind of cool tools making good use of this added information the community may come up with, but would be discouraged due the difficulty to implement.

I'm also thinking about, for instance, code highlighting on GitHub or elsewhere, which would automatically work with annotations, but won't (I think!) with special KDoc tags.

bjhham replied:
I'll look into providing extension points and exposing the compiler API here to allow for custom processing.  This way, people can use annotations or custom parsing as they see fit.

BoD commented on proposals/0004-open-api.md:
```diff
@@ -0,0 +1,322 @@
+| `@cookie`      | `@cookie name description`          | Describes a cookie parameter                                 |
+| `@body`        | `@body [Type] description`          | Documents the request body type                              |
+| `@response`    | `@response code [Type] description` | Documents a response code with optional type and description |
+| `@deprecated`  | `@deprecated reason`                | Marks an endpoint as deprecated                              |
```
Somewhat related to my other comment - do we need this when there is already the standard `@Deprecated` annotation?

bjhham replied:
You cannot use the `@Deprecated` annotation for targeting expressions, unfortunately.

bjhham commented:
Note to self: describe a mechanism for examples in schemas.

e5l reviewed (approved):
We can also think about diagnostics or inspections about API documentation on the plugin side

satamas commented on proposals/0004-open-api.md:
```diff
@@ -0,0 +1,322 @@
+ * @tag [Users]
+ * @param id The user identifier
+ * @response 200 [User] found
+ * @response 404 [User] not found
```
What about some common error codes? For example, 401 and 403 that are coming from the authorization phase. Or I may have some throttling logic that will return 429. Will it be required to specify them in every endpoint?

bjhham replied:
The KDoc parameters are merged from the stack, so you can do something like:

```kotlin
/**
 * @response 429 [ErrorMessage] Rate limit exceeded
 */
route("/api") {
    /**
     * Get a list of users.
     * @response 200 [User]+ A list of users
     */
    get("/users") {
        call.respond(repsository.list<User>())
    }
}
```

This should work with extension functions on `Route` and other blocks like `authenticated`.  I'll include more specifics on how this works.

satamas commented on proposals/0004-open-api.md:
```diff
@@ -0,0 +1,322 @@
+| `@response`    | `@response code [Type] description` | Documents a response code with optional type and description |
+| `@deprecated`  | `@deprecated reason`                | Marks an endpoint as deprecated                              |
+| `@description` | `@description text`                 | Provides a detailed endpoint description                     |
+| `@security`    | `@security scheme`                  | Documents security requirements                              |
```
What if I want to add some custom field like x-sse? Will it require some  custom comment parser function?

bjhham replied:
I'll add something for extension fields.  `@x-<field>` or something.

marychatte commented on proposals/0004-open-api.md:
```diff
@@ -0,0 +1,322 @@
+## Extensibility
+[extensibility]: #extensibility
+
+The functionality of the generation ought to be extensible in the following ways:
```
Should we support `@OpenApiIgnore` to let developers explicitly exclude endpoints from generated specs?

marychatte commented on proposals/0004-open-api.md:
```diff
@@ -0,0 +1,322 @@
+
+Note that for `@body` and `@response` fields, a type reference is specified, which will be used to automatically supply the schema definition in the OpenAPI specification.
+
+By default, all parameters will be considered required unless the name is suffixed with a question mark (`?`).  For example, `@param name?` indicates that the parameter is optional.
```
Should we include `default` as well as `optional`?
https://swagger.io/docs/specification/v3_0/describing-parameters/#default-parameter-values

osipxd replied:
In addition to this. Should we also add information about parameter type and value restrictions (range, regex, etc.)?

bjhham replied:
Yes, there's a few options to consider here.  I think we'll need some multi-line fields for when there's extra options like defaults and validation.  I'll update the doc to include these.

I also noticed that `@param` is also used with normal KDoc and it confuses the IDE because it triggers a warning, so maybe we'll need to go with `@query` / `@form` or something.

marychatte commented on proposals/0004-open-api.md:
```diff
@@ -0,0 +1,322 @@
+## Extensibility
+[extensibility]: #extensibility
+
+The functionality of the generation ought to be extensible in the following ways:
```
Maybe we can also add a strict `verifyOpenApi` check to fail the build if KDoc annotations (like `List<User>`) don’t match actual code types (e.g., `User`)? Of fail on undocumented/unreachable code too?

bjhham replied:
I was thinking it could be an option on the gradle extension like `strict = true` since the validation and generation are all rolled into the compilation process currently, where the default will be to just have warnings.

marychatte reviewed (approved):
Good 👍

osipxd reviewed (approved):
🏆

bjhham committed:
Added prototype, more field info, small changes to API

CLOVIS-AI commented:
How can users add such information to their endpoints with regular code calls?

This proposal seems very nice for the basic use-case, but configuring everything via magic tends to scale poorly. If users already have their own convenience functions over the Ktor DSL, how can they modify that DSL so it shares its information to this new system?

hfhbd commented:
How will it work with multi or even published modules?

IMHO the only options are runtime annotations or regular functions.

bjhham commented:
@hfhbd 

> How will it work with multi or even published modules?

We'll include the output in the build resources when the plugin is enabled, so it will get bundled with the module artifact.  For multiple modules, each module will need to include the gradle plugin and we should be able to stitch together these artifacts.  I'll try to include some specifics on the implementation here when I have something ready.

> IMHO the only options are runtime annotations or regular functions.

We'll investigate a more robust runtime solution using a new routing DSL, but runtime annotations won't work for expressions (i.e., routing calls).  Our focus for this iteration is to provide something that can work with existing projects using the current routing DSL, so that no code changes are required for adoption.

bjhham commented:
@CLOVIS-AI I don't have specifics yet, but I'll try to provide more details on calls to include information in the resulting model.

bjhham committed:
Minor updates; added @ignore

bjhham committed:
Fix typos

bjhham committed:
Updates regarding compiler plugin and runtime access

bjhham committed:
Fixup table of contents

ArtyomGabeev commented:
Hi there, thanks for the great work!

Will it be possible to provide examples of request/response bodies? May be we can reference them from classpath resources?

whyoleg commented on proposals/0004-open-api.md:
````diff
@@ -0,0 +1,446 @@
+}
+```
+
+### KDoc Fields
````
FYI: those are called `tags` in both KDoc and Javadoc.
Javadoc supports custom tags, while custom tags in KDoc are _somehow_ supported only in Dokka.

From this standpoint, I would strongly suggest avoiding the `@xxx` syntax, as it will only complicate the understanding of KDoc (in both ktor and on declarations) and may eventually require us to support such custom tags _somehow_ in compiler-related APIs.

WDYT about using some other markdown syntax for this? Especially, as you are using additional custom syntax for references...

Lately, the Dokka and Analysis API team has been working hard on the KDoc link resolution algorithm improvements ([KEEP#0385](https://github.com/Kotlin/KEEP/blob/main/proposals/KEEP-0385-kdoc-links-to-extensions.md) and [KEEP#0389](https://github.com/Kotlin/KEEP/blob/main/proposals/KEEP-0389-kdoc-streamline-ambiguous-KDoc-links.md)), but, as far as I understand, you are not using Analysis API (which is the source of truth for KDoc link resolution), and so we will have the same syntax in KDoc which will work significantly different...

We would also like to continue standardizing all other work related to KDoc link resolution, tags, specifications, and so on.

bjhham replied:
That's a fair point.  We can change it to use more of a YAML-like syntax like `foo: bar` for attributes since it would be familiar for OpenAPI users.  The only reason for using KDoc tag syntax was that it was how this was first imagined and I had no compelling reason to avoid it.

whyoleg replied:
Yes, or, you could just use markdown :)

I mean, like this:
```kotlin
/**
 * Eat the provided specialty sloth food.
 *
 * - Parameters:
 *    - food: The food for the sloth to eat.
 *    - quantity: The quantity of the food for the sloth to eat.
 * - Returns: The sloth's energy level after eating.
 */
```
IDE can provide autocomplete for those predefined `Parameters`/`Returns` sections.

P.S. This specific suggestion/syntax is inspired by [Swift](https://developer.apple.com/documentation/xcode/writing-symbol-documentation-in-your-source-files#Describe-the-parameters-of-a-method) :)
P.S.S. Javadoc also has something similar mentioned in [JEP 467](https://openjdk.org/jeps/467) (under `Future Work`):
```md
# Parameters

* x   the x coordinate
* y   the y coordinate
```

At the moment, there are no plans to support something like this in KDoc. :)

bjhham replied:
Yeah, that would be a more consistent experience.  Generating Kotlin from Markdown feels a bit strange, but it would be nice and readable as documentation.

whyoleg commented on proposals/0004-open-api.md:
````diff
@@ -0,0 +1,446 @@
+From the routing DSL, the inclusion of path information would appear as:
+
+```kotlin
+routing {
+    get("/articles") {
+        val query = call.queryParameters["q"]?.let(::parseQuery)
+        call.respond(articleRepository.findArticles(query))
+    }.annotate {
+        parameters {
+            query("q")
+        }
+        responses {
+            HttpStatusCode.OK {
+                summary = "A list of articles"
+                contentType = ContentType.Application.Json
+                schema = jsonSchema<List<Article>>()
+            }
+        }
+    }
+}
````
What if? Or why not?

```suggestion
routing {
    get("/articles", documentation = {
        parameters {
            query("q")
        }
        responses {
            HttpStatusCode.OK {
                summary = "A list of articles"
                contentType = ContentType.Application.Json
                schema = jsonSchema<List<Article>>()
            }
        }
    }) {
        val query = call.queryParameters["q"]?.let(::parseQuery)
        call.respond(articleRepository.findArticles(query))
    }
}
```

Yes, it will force us to change the current API, but it seems possible to do so in a backward-compatible manner (binary and source-wise).

bjhham replied:
There's no absolute blocker there, but there's a few reasons that make it somewhat impractical...

1. There are a lot of routing functions we'd need to modify; the routing API is deceptively large.
2. We don't want to introduce OpenAPI into the core library and keep it as an optional extension.
3. It's a bit cleaner to chain function calls for the generated code, so there's clear boundaries and scopes for when there includes both the generated documentation call and the human-made call `get {}.annotate {}.annotate {}`.  With the single lambda, we'd have to add the expressions onto the end.
4. We're mixing concerns in the routing calls here with documentation and route filtering.  It could look a bit confusing if you have path parameters in the path and the same definitions in the documentation as well.

We're also considering developing a new routing API that has all these details built in where you can describe everything in one go, but imo it would need a kind of different signature.  @nomisRev has a project with the general concept https://github.com/nomisRev/ktor-typed though we haven't worked out the details.

whyoleg replied:
Agree, those are valid concerns!
It would be helpful to mention the reasoning behind the current `annotate` API, as it is mentioned in the document (sorry if it's mentioned elsewhere and I just missed it).

As for me, I don't have a problem with the `annotate` API being specified after `get{}`.

Minor/Major?: I feel that `get {}.annotate {}.annotate {}` should not be allowed, as the merging strategy here will be too complex, especially in the case of lists. For example, if we have multiple `annotate {}.annotate {}`, should we merge or override the `responses` configuration? What about `parameters`? Should we merge them by parameter name or replace the lists? And so on...
I believe it's possible to define a good merging strategy, but to me, it feels that there should be only one source of truth, especially if we allow it to be done both via code (`annotate {}`) and via the KDoc analyzer at the same time.

bjhham replied:
I think the only reasonable alternative atm would be something like:

```kotlin
annotate {
    parameters {
        query("q")
    }
    responses {
        HttpStatusCode.OK {
            summary = "A list of articles"
            contentType = ContentType.Application.Json
            schema = jsonSchema<List<Article>>()
        }
    }
}.get("/articles") {
    val articles = repository.list()
    call.respond(articles)
}
```

Which has the correct chronology of `describe -> implement`, but it would be a bit weird for the actual execution, because the `annotate` call creates a wrapper routing context that just injects the metadata after the first route is created from it, whereas with the postfix call it's just assigning the metadata on the route that's returned from `get`.  Maybe if the prefix call was a greatly superior developer experience it would be worth changing, though I'm not convinced it would make that much of a difference...

bjhham replied:
The merging strategy I think is pretty crucial when you're working with the generated code, so the general approach is that the last one wins and lists are concatenated where applicable.

bjhham committed:
Modified commenting spec to use markdown keywords instead of KDoc tags

Informed by @whyoleg's comment regarding potential problems with using KDoc tags

bjhham committed:
Update naming and gradle plugin details
